test/test_type_promotion.py

```diff
@@ -0,0 +1,201 @@
+        self.assertRaisesRegex(RuntimeError, "can't be cast to", lambda: uint8_tensor.add_(half_tensor))
+
+        # We allow demotion from signed to unsigned, unlike numpy, because we
+        # don't want the performance penalty of inspecting scalar values, and
+        # because we don't want 'signed' to be considered a distinct 'category'
+        # in promotion rules.
+        # If signed were a distinct category, uint16_tensor + 5 would result in
+        # a long_tensor, which is not what we want.
+        int16_tensor = torch.tensor([1, 1, 1], dtype=torch.int16)
+        uint8_tensor *= int16_tensor
+
+    def test_unsinged(self):
+        dont_promote = torch.ones(3, dtype=torch.uint8) + 5
+        self.assertEqual(dont_promote.dtype, torch.uint8)
+
+    # some basic examples
+
+    def test_int_promotion(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.int64, device=self.device)
+        c = a + b
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.int64)
+
+    def test_float_promotion(self):
+        a = torch.ones([4, 4, 4], dtype=torch.float, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.double, device=self.device)
+        c = a + b
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.double)
+        c = b + a
+        self.assertEqual(c, b + b)
+        self.assertEqual(c.dtype, torch.double)
+
+    def test_add_wrapped(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int, device=self.device)
+        b = 1
+        c = a + b
+        self.assertEqual(c, a + a)
+        self.assertEqual(c.dtype, torch.int)
+
+    def test_int_to_float(self):
+        a = torch.ones([4, 4, 4], dtype=torch.int32, device=self.device)
+        b = torch.ones([4, 4, 4], dtype=torch.float, device=self.device)
+        c = a + b
+        self.assertEqual(c.dtype, torch.float32)
+
+    # some examples from:
+    # https://github.com/pytorch/pytorch/issues/9515
+
+    def test_from_issue(self):
+        a = torch.rand(3, dtype=torch.float32, device=self.device)
+        u = torch.tensor([0, 0, 1], dtype=torch.uint8, device=self.device)
+        self.assertEqual((a * 5).dtype, torch.float32)
+        self.assertEqual((u + 1).dtype, torch.uint8)
+        self.assertEqual((u + 1000).dtype, torch.uint8)  # integer overflow
+
+        # not a "wrapped number"
+        other = torch.tensor(5.5, dtype=torch.double, device=self.device)
+
+        self.assertEqual((u + 5.5).dtype, torch.get_default_dtype())
+        self.assertEqual((u + other).dtype, torch.double)
+        # adding a 0-dim tensor to a float doesn't promote to double unless first
+        # type was integral.
+        self.assertEqual((a + other).dtype, torch.float32)
+
+    def test_half(self):
+        if(self.device == 'cpu'):
+            raise unittest.SkipTest('add_cpu not implemented for Half')
```

Comment:
we shouldn't be skipping test except if it's something conditionally compiled (so it can't actually run the code in question).  The problem with this is we'll never actually run this code, even when we enable add for Half.  Instead, assert that this raises an exception so when we implement addition, someone will ensure the test runs.